### ISSUE TITLE
Assigned fill to selected bar using 'cell' disappears when you move the brush

### ISSUE THREAD

**Issue by wonjiky** [recharts/recharts#1513]
Hi, 

I'm using the bar chart with brush. The selected bar is assigned a different fill using the <Cell/> method but whenever I move zoom in with the brush, the highlighted attribute is no longer in effect. It is literally only effective when I don't use the brush... If anyone could give me an insight or any kind of explanation on what is going on, I would greatly appreciate it.

This is the line of my <Cell />: 
` {data.map((entry, index) => (
                        <Cell cursor="pointer" key={`cell-${index}`} fill={entry.ID ===  selectedAgglos ? '#c95d47' : 'lightgrey' }/>
                    ))}`

Thank you

**Comment by wonjiky** [comment]
I realized the <Cell /> is only applicable for bars that are currently rendered? So when I assign a single bar to have a different fill attribute, it does some weird action. For example, just randomly assigns the fill attribute to another bar or when zoomed in, the fill attribute is no longer applied...

Is there any other way to assign a bar to with a different fill attribute based on external selection while using the brush interaction ?.. :(

**Comment by marco-streng** [comment]
Same problem here. 

I've made a simple example: https://jsfiddle.net/2d9ncbf6/
When you drag the left end from the brush, the second bar get's the color (blue) of the first bar and so on ...

**Comment by zemuldo** [comment]
I have the same problem.

**Comment by tigran-sedrakyan** [comment]
Hey there! Seems like the problem is caused by recharts keeping the original Bar fill array, despite the change in the displayed Bar array. Came up with a hacky workaround, by keeping the startIndex of the brush in the state and changing the fill values for the Bars depending on the startIndex. Here's a little modification over @marco-streng 's example: 

[(https://codesandbox.io/s/52k9k2pkmx)]

**Comment by gndelia** [comment]
I'm facing a similar issue. I will try the workaround, but the bug persists

**Comment by gndelia** [comment]
A workaround I was able to apply, is to use multiple axis and multiple bars, but hide all of the X-axis but one. Each bar has a different fill color, and has a different `dataKey`. So what I do to use that configuration is to have a property per type of value, meaning that instead of defining the color in `<Cell />`, I just use different properties (Which implies mapping the data before passing it into the chart) for each expected color, and then I use one bar per color. That worked.

**Comment by shahinlotfi** [comment]
I was having the same issue. Here's how I solved it, using a custom shape.

```
  const CustomBar = (props) => {
    const { vibration } = props

    // Don't forget to import Rectangle

    return (
      <Rectangle
        {...props}
        fill={
          vibration <= 25  ? '#FF0000' : '#00FF00'
        }
      />
    )
  }
```

Usage:
``` 
<Bar shape={CustomBar} dataKey="vibration" stackId="a" name="Vibration" />
````

**Comment by Andyyyyyy** [comment]
I stumbled across this problem myself, in my case I wanted to have one particular bar in a different color.

I managed to solve this by keeping track of the `startIndex` and then calculating the correct index when the brush changes. 
`index === activeIndex - startIndex`

This might be trivial but could save some time for anyone who also comes across this issue.

https://codesandbox.io/s/recharts-brush-with-custom-cell-fix-4ck18?file=/src/App.tsx

**Comment by ariel-upstream** [comment]
Why was this issue closed? The bug still persists

**Comment by ckifer** [comment]
I believe this will be fixed in 3.0 but let me confirm

**Comment by ckifer** [comment]
I mean _technically_ in the sense that you are assigning a specific index to a specific color, recharts is behaving correctly. 

You slice off 1 value from an array of data and the new index for `array[1]` is `array.length - 1`

**Comment by PavelVanecek** [comment]
Yeah agreed that this is _technically_ the correct behavior because the `Cell` elements follow the indexes of data and if the data index changes then you should also update how many `Cell` components you render.

I also agree that this is a horrible design and we should change it! I am planning to propose an alternative approach to using `Cell` and then releasing it in a minor version. Stay tuned for more updates.

**Comment by chwiseena311** [comment]
Hi, any update on this?
